design-documents/stateless-api-token.md

```diff
@@ -0,0 +1,147 @@
+ 
+ 
+### Overview of new stateless tokens implementation
+* `TokenUserContext` to be updated to decode new tokens and use old logic for existing tokens
+* [Create and implement](#spi-for-token-data-manager) SPIs for providing user data to be encoded in tokens and reading it
+* Create and [implement](#encoding-and-signing-tokens) SPI for singing and encoding token data
+* Update token revoke functionality to use cache storage to store revoke request instead of DB for customer/admin tokens
+  [details](#how-to-handle-manual-expiration)
+ 
+ 
+### Encoding and signing tokens
+Tokens must have enough data encoded in order to authenticate a user. At minimum it would contain:
+* user type
+* user ID
+* issued timestamp
+* expires timestamp (in order to be able to validate token without knowing token lifespan)
+ 
+3rd-party developers will be able to extend this list via [the SPI](#spi-for-token-data-manager).
+ 
+After serialization this data must be signed to confirm origin. Magento instance(es) will have the same secret used
+to decode the tokens without having to validate them via DB.
+ 
+Magento will have _SPI_ that will handle encoding/decoding tokens and allow 3rd-party developers to change
+how tokens are handled:
+ 
+```php
+interface AuthData
+{
+    public function getIssued(): int;
+
+    public function getExpires(): int;
+
+    public function getUserData(): array;
+}
+
+
+interface TokenEncoderInterface
+{
+    public function encode(AuthData $data): string;
+
+    public function decode(string $token): AuthData;
+}
+``` 
+ 
+There are 2 ways we can go about implementing this:
+ 
+#### Own implementation using EncyptorInterface
+User data returned from [the SPI](#spi-for-token-data-manager) in form of array of scalars will be serialized into string, paired with
+signature, base64 encoded for obfuscation after.
+ 
+Example in pseudo-code:
+```php
+$userData = ['user_data' => ['user_id' => 1, 'user_type' => 3], 'issued' => time(), 'expires' => time() + (60*60)];
+$encodedData = json_encode($userData);
+$signature = $encryptor->hash($encodedData, true);
+
+return $token = 'token:' .base64_encode($encodedData) .'.' .base64_encode($signature);
+```
+ 
+__Pros__:
+* Easy to implement
+* Uses EncryptorInterface which guarantees up-to-date secure encryption
+* No new project dependencies introduced
+ 
+__Cons__:
+* Custom solution
+ 
+#### Using JWT
+JWT is an established standard for authentication tokens that contain user data and are signed as origin verification.
+A number of [PHP libraries](https://jwt.io) existing implementing it.
+ 
+__Pros__:
+* Established standard
+* Existing libraries
+* Wide variety of supported encryption algorithms
+* Security concerns resolved by standard and libraries themselves
+ 
+__Cons__:
+* Will have to add new project dependency for a library and it's dependencies
+* Some libraries have [known security vulnerabilities](https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/)
```

Comment:
It's easy to update 3rd party library when finding a security problem in the custom implementation. Also, such tools as Github track security vulnerabilities in the packages.

Reply:
it is easy, but it takes time due to our release strategy and how long it takes an issue to reach core devs